File: .yalc/@entur/atb-mobile-client-sdk/challenge/bluetooth/model/bluetooth-model-java-lib/src/main/java/no/entur/abt/android/bluetooth/header/ChallengeAdvertisingFrameHeaderBinaryFormat.java
package no.entur.abt.android.bluetooth.header;

/**
 * Header format <br>
 * <br>
 * - Channel (3 bit)<br>
 * - Last frame (1 bit)<br>
 * - Frame number (4 bit)<br>
 * - Client id (2 bytes)<br>
 * - Correlation id (1 byte)<br>
 * <br>
 * So in total 4 bytes.
 */

public class ChallengeAdvertisingFrameHeaderBinaryFormat {

	public static final int FRAME_HEADER_LENGTH = 4;

	private static final int LAST_FRAME_MASK = 0b00010000;
	private static final int CHANNEL_MASK = 0b00000111;
	private static final int FRAME_NUMBER_MASK = 0b00001111;

	public static final int MAX_FRAMES = 16;
	private static final int MAX_CLIENT_ID = 65535;
	private static final int MAX_CORRELATION_ID = 255;

	public ChallengeAdvertisingFrameHeader parse(byte[] headerBytes) {
		byte b = headerBytes[0]; // CCCLNNNN

		int channel = (b >>> 5) & CHANNEL_MASK; // CCC
		boolean lastFrame = (b & LAST_FRAME_MASK) != 0; // L
		int frameNumber = b & FRAME_NUMBER_MASK; // NNNN

		// see DataInputStream.readUnsignedShort()
		int clientId = ((headerBytes[1] & 0xFF) << 8) + (headerBytes[2] & 0xFF);
		// see DataInputStream.readUnsignedByte()
		int correlationId = headerBytes[3] & 0xFF;

		return new ChallengeAdvertisingFrameHeader(clientId, frameNumber, lastFrame, correlationId, channel);
	}

	public byte[] serialize(ChallengeAdvertisingFrameHeader header) {
		byte[] buffer = new byte[FRAME_HEADER_LENGTH];

		return serialize(buffer, header);
	}

	public byte[] serialize(byte[] buffer, ChallengeAdvertisingFrameHeader header) {
		int frameNumber = header.getFrameNumber();
		if (frameNumber >= MAX_FRAMES || frameNumber < 0) {
			throw new IllegalArgumentException("Frame number must be 0-" + (MAX_FRAMES - 1) + ", got " + frameNumber);
		}
		int channel = header.getChannel();
		if (channel > 7 || channel < 0) {
			throw new IllegalArgumentException("Channel must be 0-7, got" + channel);
		}
		int correlationId = header.getCorrelationId();
		if (correlationId > MAX_CORRELATION_ID || correlationId < 0) {
			throw new IllegalArgumentException("Correlation id must be 0-" + MAX_CORRELATION_ID + ", got" + channel);
		}
		int clientId = header.getClientId();
		if (clientId > MAX_CLIENT_ID || clientId < 0) {
			throw new IllegalArgumentException("Client id must be 0-" + MAX_CLIENT_ID + ", got" + channel);
		}

		int h = channel << 5 | frameNumber;
		if (header.isLast()) {
			h = h | LAST_FRAME_MASK;
		}
		buffer[0] = (byte) (h & 0xFF);

		// see DataOutputStream.writeShort()
		buffer[1] = (byte) ((clientId >>> 8) & 0xFF);
		buffer[2] = (byte) ((clientId >>> 0) & 0xFF);
		buffer[3] = (byte) (correlationId & 0xFF);

		return buffer;
	}

	public int getLength() {
		return FRAME_HEADER_LENGTH;
	}

}

File: .yalc/@entur/atb-mobile-client-sdk/challenge/bluetooth/model/bluetooth-model-ios-lib/Sources/bluetooth-model-ios-lib/header/ChallengeAdvertisingFrameHeaderBinaryFormat.swift

import Foundation

enum FrameHeaderBinaryFormatError: Error {
    case rangeError(String)
}

/**
 * Header format <br>
 * <br>
 * - Channel (3 bit)<br>
 * - Last frame (1 bit)<br>
 * - Frame number (4 bit)<br>
 * - Client id (2 bytes)<br>
 * - Correlation id (1 byte)<br>
 * <br>
 * So in total 4 bytes.
 */

class ChallengeAdvertisingFrameHeaderBinaryFormat {
    
    public static let FRAME_HEADER_LENGTH = 4
    
    static let LAST_FRAME_MASK : UInt8 = 0b00010000
    static let CHANNEL_MASK : UInt8 = 0b00000111
    static let FRAME_NUMBER_MASK : UInt8 = 0b0001111
    
    static let MAX_FRAMES = 16
    static let MAX_CLIENT_ID = 65535
    static let MAX_CORRELATION_ID = 255
    
    public func parse(headerBytes: inout Data) -> ChallengeAdvertisingFrameHeader {
        let b = headerBytes[0]
        
        let channel : UInt32 = UInt32((b >> 5) & ChallengeAdvertisingFrameHeaderBinaryFormat.CHANNEL_MASK)
        
        let lastFrame = (b & ChallengeAdvertisingFrameHeaderBinaryFormat.LAST_FRAME_MASK) != 0
        
        let frameNumber : UInt32 = UInt32(b & ChallengeAdvertisingFrameHeaderBinaryFormat.FRAME_NUMBER_MASK)
        
        let clientId : UInt32 = UInt32(((headerBytes[1] << 8) + headerBytes[2]))
        
        let correlationId : UInt32 = UInt32(headerBytes[3])
        
        return ChallengeAdvertisingFrameHeader(clientId: clientId, frameNumber: frameNumber, lastFrame: lastFrame, correlationId: correlationId, channel: channel)
    }
    
    public func serialize(header: ChallengeAdvertisingFrameHeader) throws -> Data {
        var buffer = Data()
        try serialize(buffer: &buffer, header: header)
        return buffer
    }
    
    public func serialize(buffer : inout Data, header: ChallengeAdvertisingFrameHeader) throws {
        let frameNumber = header.frameNumber
        if(frameNumber >= ChallengeAdvertisingFrameHeaderBinaryFormat.MAX_FRAMES) {
            throw FrameHeaderBinaryFormatError.rangeError("Frame number must be 0-\(ChallengeAdvertisingFrameHeaderBinaryFormat.MAX_FRAMES - 1), got \(frameNumber)")
        }
        let channel = header.channel
        if(channel > 7) {
            throw FrameHeaderBinaryFormatError.rangeError("Channel number must be 0-7, got \(channel)")
        }
        let correlationId = header.correlationId
        if(correlationId > ChallengeAdvertisingFrameHeaderBinaryFormat.MAX_CORRELATION_ID) {
            throw FrameHeaderBinaryFormatError.rangeError("Correlation id must be 0-\(ChallengeAdvertisingFrameHeaderBinaryFormat.MAX_CORRELATION_ID), got \(correlationId)")
        }
        let clientId = header.clientId
        if(clientId > ChallengeAdvertisingFrameHeaderBinaryFormat.MAX_CLIENT_ID) {
            throw FrameHeaderBinaryFormatError.rangeError("Client id must be 0-\(ChallengeAdvertisingFrameHeaderBinaryFormat.MAX_CLIENT_ID), got \(clientId)")
        }

        var h : UInt8 = UInt8(channel << 5 | frameNumber)
        if(header.lastFrame) {
            h = h | ChallengeAdvertisingFrameHeaderBinaryFormat.LAST_FRAME_MASK
        }
        buffer.append(h)
        buffer.append( UInt8( (clientId >> 8) & 0xFF) )
        buffer.append( UInt8( (clientId >> 0) & 0xFF) )
        buffer.append( UInt8( correlationId) )
    }

    public func getLength() -> Int {
        return ChallengeAdvertisingFrameHeaderBinaryFormat.FRAME_HEADER_LENGTH
    }
}
